pallets/xvm/src/lib.rs

```diff
@@ -0,0 +1,278 @@
+// along with Astar. If not, see <http://www.gnu.org/licenses/>.
+
+//! # XVM pallet
+//!
+//! A module to provide
+//!
+//! ## Overview
+//!
+//! The XVM pallet provides a runtime interface to call different VMs. It currently
+//! supports two VMs: EVM and WASM. With further development, more VMs can be added.
+//!
+//! Together with other functionalities like Chain Extension and precompiles,
+//! the XVM pallet enables the runtime to support cross-VM calls.
+//!
+//! ## Interface
+//!
+//! ### Implementation
+//!
+//! - Implements `XvmCall` trait.
+//!
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{ensure, traits::ConstU32, BoundedVec};
+use pallet_contracts::{CollectEvents, DebugInfo, Determinism};
+use pallet_evm::GasWeightMapping;
+use parity_scale_codec::Decode;
+use sp_core::U256;
+use sp_runtime::traits::StaticLookup;
+use sp_std::{marker::PhantomData, prelude::*};
+
+use astar_primitives::{
+    ethereum_checked::{
+        AccountMapping, CheckedEthereumTransact, CheckedEthereumTx, MAX_ETHEREUM_TX_INPUT_SIZE,
+    },
+    xvm::{CallError, CallErrorWithWeight, CallInfo, CallResult, Context, VmId, XvmCall},
+};
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+pub mod weights;
+pub use weights::WeightInfo;
+
+#[cfg(test)]
+mod mock;
+
+pub use pallet::*;
+
+pub type WeightInfoOf<T> = <T as Config>::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config + pallet_contracts::Config {
+        /// Mapping from `Account` to `H160`.
+        type AccountMapping: AccountMapping<Self::AccountId>;
+
+        /// Mapping from Ethereum gas to Substrate weight.
+        type GasWeightMapping: GasWeightMapping;
+
+        /// `CheckedEthereumTransact` implementation.
+        type EthereumTransact: CheckedEthereumTransact;
+
+        /// Weight information for extrinsics in this pallet.
+        type WeightInfo: WeightInfo;
+    }
+}
+
+impl<T: Config> XvmCall<T::AccountId> for Pallet<T> {
+    fn call(
+        context: Context,
+        vm_id: VmId,
+        source: T::AccountId,
+        target: Vec<u8>,
+        input: Vec<u8>,
+    ) -> CallResult {
+        Pallet::<T>::do_call(context, vm_id, source, target, input, false)
+    }
+}
+
+impl<T: Config> Pallet<T> {
+    fn do_call(
+        context: Context,
+        vm_id: VmId,
+        source: T::AccountId,
+        target: Vec<u8>,
+        input: Vec<u8>,
+        skip_execution: bool,
+    ) -> CallResult {
+        ensure!(
+            context.source_vm_id != vm_id,
+            CallErrorWithWeight {
+                error: CallError::SameVmCallNotAllowed,
+                used_weight: match vm_id {
+                    VmId::Evm => WeightInfoOf::<T>::evm_call_overheads(),
+                    VmId::Wasm => WeightInfoOf::<T>::wasm_call_overheads(),
+                },
+            }
+        );
+
```

Comment:
About re-entrancy, I think both are solved in their respective VMs/crates.

EVM takes care of it [here](https://github.com/AstarNetwork/frontier/blob/polkadot-v0.9.43/frame/evm/src/runner/stack.rs#L56).
WASM takes care of it [here](https://github.com/paritytech/substrate/pull/13443).

Perhaps we should also prohibit re-entrancy into XVM itself?

Reply:
Created an issue for it. Will do follow-up after this PR. https://github.com/AstarNetwork/Astar/issues/986